File: homework01-go/caesar/caesar.go
package caesar

func EncryptCaesar(plaintext string, shift int) string {
	var ciphertext string
	var language int = 26

	for i := 0; i < len(plaintext); i++ {
		sym_num := int(plaintext[i])
		if (int('A') <= sym_num) && (sym_num <= int('Z')) {
			sym_num = (int(sym_num)+shift-int('A'))%language + int('A')
		} else if (int('a') <= sym_num) && (sym_num <= int('z')) {
			sym_num = (int(sym_num)+shift-int('a'))%language + int('a')
		}
		ciphertext += string(rune(sym_num))
	}
	return ciphertext
}

func DecryptCaesar(ciphertext string, shift int) string {
	var plaintext string
	var language int = 26

	for i := 0; i < len(ciphertext); i++ {
		sym_num := int(ciphertext[i])
		if (int('A') <= sym_num) && (sym_num <= int('Z')) {
			sym_num = (sym_num-shift-int('A')+language)%language + int('A')
		} else if (int('a') <= sym_num) && (sym_num <= int('z')) {
			sym_num = (sym_num-shift-int('a')+language)%language + int('a')
		}
		plaintext += string(rune(sym_num))
	}
	return plaintext
}

File: homework01/caesar.py
import typing as tp


def encrypt_caesar(plaintext: str, shift: int = 3) -> str:
    """
    Encrypts plaintext using a Caesar cipher.

    >>> encrypt_caesar("PYTHON")
    'SBWKRQ'
    >>> encrypt_caesar("python")
    'sbwkrq'
    >>> encrypt_caesar("Python3.6")
    'Sbwkrq3.6'
    >>> encrypt_caesar("")
    ''
    """
    ciphertext = ""
    a_first = ord("a")
    z_last = ord("z")
    A_first = ord("A")
    Z_last = ord("Z")
    language = 26
    for i in range(len(plaintext)):
        code_letter = ord(plaintext[i])
        if a_first <= code_letter <= z_last:
            k = chr((code_letter - a_first + shift) % language + a_first)
            ciphertext += k
        elif A_first <= code_letter <= Z_last:
            k = chr((code_letter - A_first + shift) % language + A_first)
            ciphertext += k
        else:
            ciphertext += plaintext[i]

    return ciphertext


def decrypt_caesar(ciphertext: str, shift: int = 3) -> str:
    """
    Decrypts a ciphertext using a Caesar cipher.

    >>> decrypt_caesar("SBWKRQ")
    'PYTHON'
    >>> decrypt_caesar("sbwkrq")
    'python'
    >>> decrypt_caesar("Sbwkrq3.6")
    'Python3.6'
    >>> decrypt_caesar("")
    ''
    """
    plaintext = ""
    a_first = ord("a")
    z_last = ord("z")
    A_first = ord("A")
    Z_last = ord("Z")
    language = 26
    for i in range(len(ciphertext)):
        code_letter = ord(ciphertext[i])
        if a_first <= code_letter <= z_last:
            check_negative = code_letter - a_first - shift
            if check_negative < 0:
                check_negative = (
                    language - (check_negative * (-1) % language)
                ) % language
            k = chr(check_negative + a_first)
            plaintext += k
        elif A_first <= code_letter <= Z_last:
            check_negative = code_letter - A_first - shift
            if check_negative < 0:
                check_negative = (
                    language - (check_negative * (-1) % language)
                ) % language
            k = chr(check_negative + A_first)
            plaintext += k
        else:
            plaintext += ciphertext[i]

    return plaintext


def caesar_breaker_brute_force(ciphertext: str, dictionary: tp.Set[str]) -> int:
    """
    Brute force breaking a Caesar cipher.
    """
    best_shift = 0
    # PUT YOUR CODE HERE
    return best_shift
